contracts/javascore/ibc/src/main/java/ibc/ics20/app/ICS20Transfer.java

```diff
@@ -0,0 +1,179 @@
+package ibc.ics20.app;
+
+import ibc.icon.interfaces.IIBCModule;
+import ibc.icon.score.util.Logger;
+import ibc.icon.score.util.StringUtil;
+import ibc.ics05.port.ModuleManager;
+import ibc.ics25.handler.IBCHandler;
+
+import icon.proto.core.channel.Channel;
+import icon.proto.core.channel.Packet;
+import score.Address;
+import score.Context;
+import score.DictDB;
+import score.annotation.External;
+
+import java.math.BigInteger;
+import java.util.Map;
+
+public class ICS20Transfer {
+    public static String ICS20_VERSION = "ics20-1";
+    public static Address ZERO_ADDRESS = Address.fromString("hx0000000000000000000000000000000000000000");
+
+
+    public static final DictDB<String, Address> channelEscrowAddresses = Context.newDictDB("channelEscrowAddresses", Address.class);
+
+
+    @External
+    public byte[] onRecvPacket(byte[] packet, Address relayer) {
+//        TODO unmarshal json
+        boolean success = false;
+        Address receiver = ZERO_ADDRESS;
+        receiver, success = _decodeReceiver(packet.data.receiver);
+        if (!success) {
+           return ICS20Lib.FAILED_ACKNOWLEDGEMENT_JSON;
+        }
+
+        byte[] denomPrefix = getDenomPrefix(packet.data.sourcePort, packet.data.sourceChannel);
+        byte[] denom = packet.data.denom.getBytes();
+        if (denom.length >= denomPrefix.length) {
+//            denom slicing todo
+            success = _transferFrom(getEscrowAddress(packet.data.sourceChannel), receiver, packet.data.denom, packet.data.amount);
+        } else{
+            if(ICS20Lib.isEscapeNeededString(denom)){
+                success = false;
+            }else{
+                success = _mint(receiver, String(getDenomPrefix(packet.destination_port, packet.destination_channel), denom), packet.data.amount);
+            }
+        }
+
+        if (success) {
+            return ICS20Lib.SUCCESSFUL_ACKNOWLEDGEMENT_JSON;
+        }else{
+            return ICS20Lib.FAILED_ACKNOWLEDGEMENT_JSON;
+        }
+
+
+        return packet;
+    }
+
+
+    @External
+    public void onAcknowledgementPacket(byte[] calldata, byte[] acknowledgement, Address relayer) {
+        Context.println("onAcknowledgementPacket");
+//        Context.require(Context.getCaller().equals(getIBCAddress()), "caller is not handler");
```

Comment:
we can decode packet from calldata ... and lets rename it to packetBytes as well.